common/scala/src/main/scala/whisk/core/entity/ActivationResult.scala

```diff
@@ -88,29 +88,50 @@ protected[core] object ActivationResponse extends DefaultJsonProtocol {
     protected[core] def payloadPlaceholder(payload: Option[JsObject]) = ActivationResponse(ApplicationError, payload)
 
     /**
+     * Class of errors for invoker-container communication.
+     */
+    protected[core] sealed abstract class ContainerConnectionError
+    protected[core] case class NoHost() extends ContainerConnectionError
+    protected[core] case class ConnectionError(t: Throwable) extends ContainerConnectionError
+    protected[core] case class NoResponseReceived() extends ContainerConnectionError
+    protected[core] case class Timeout() extends ContainerConnectionError
+    /**
+     * @param code the http status code
+     * @param entity the entity response as string
+     * @param truncated either None to indicate complete entity or Some(actual length, max allowed)
+     */
+    protected[core] case class ContainerResponse(code: Int, entity: String, truncated: Option[(ByteSize, ByteSize)] = None) {
+        val ok = code == 200 && truncated.isEmpty
+    }
+
+    /**
      * Interprets response from container after initialization. This method is only called when the initialization failed.
      *
-     * @param response an Option (HTTP Status Code, HTTP response bytes as String)
+     * @param response an either a container error or container response (HTTP Status Code, HTTP response bytes as String)
      * @return appropriate ActivationResponse representing initialization error
      */
-    protected[core] def processInitResponseContent(response: Option[(Int, String)], logger: Logging): ActivationResponse = {
-        require(response map { _._1 != 200 } getOrElse true, s"should not interpret init response when status code is 200")
-
-        response map {
-            case (code, contents) =>
-                logger.debug(this, s"init response: '$contents'")
-                Try { contents.parseJson.asJsObject } match {
-                    case scala.util.Success(result @ JsObject(fields)) =>
-                        // If the response is a JSON object container an error field, accept it as the response error.
-                        val errorOpt = fields.get(ERROR_FIELD)
-                        val errorContent = errorOpt getOrElse invalidInitResponse(contents).toJson
-                        containerError(errorContent)
-                    case _ =>
-                        containerError(invalidInitResponse(contents))
-                }
-        } getOrElse {
-            // This indicates a terminal failure in the container (it exited prematurely).
-            containerError(abnormalInitialization)
+    protected[core] def processInitResponseContent(response: Either[ContainerConnectionError, ContainerResponse], logger: Logging): ActivationResponse = {
+        require(response.isLeft || !response.right.get.ok, s"should not interpret init response when status code is OK")
```

Comment:
same as below, could be

`require(response.isLeft || !response.right.exists(_.ok), ...)`

I'm aware that it's safe to call `get` here. I'm just avoiding it in general.